vowpalwabbit/automl.cc

```diff
@@ -0,0 +1,930 @@
+               .keep()
+               .default_value("none")
+               .help("set function to determine next config {none, least_exclusion}"))
+      .add(make_option("priority_challengers", priority_challengers)
+               .keep()
+               .default_value(-1)
+               .help("set number of priority challengers to use"));
+
+  if (!options.add_parse_and_check_necessary(new_options)) return nullptr;
+
+  float (*calc_priority)(const exclusion_config&, const std::map<namespace_index, size_t>&);
+
+  if (priority_type == "none") { calc_priority = &calc_priority_empty; }
+  else if (priority_type == "least_exclusion")
+  {
+    calc_priority = &calc_priority_least_exclusion;
+  }
+  else
+  {
+    THROW("Error: Invalid priority function provided");
+  }
+
+  if (priority_challengers < 0) { priority_challengers = (static_cast<int>(max_live_configs) - 1) / 2; }
+
+  auto cm = VW::make_unique<interaction_config_manager>(
+      global_lease, max_live_configs, all.random_seed, static_cast<size_t>(priority_challengers), calc_priority);
+  auto data = VW::make_unique<automl<interaction_config_manager>>(std::move(cm));
+  // all is not needed but good to have for testing purposes
+  data->all = &all;
+  assert(max_live_configs <= MAX_CONFIGS);
+
+  // override and clear all the global interactions
+  // see parser.cc line 740
+  all.interactions.clear();
+  assert(all.interactions.empty() == true);
+
+  // make sure we setup the rest of the stack with cleared interactions
+  // to make sure there are not subtle bugs
+  auto* base_learner = stack_builder.setup_base_learner();
+
+  assert(all.interactions.empty() == true);
+
+  assert(all.weights.sparse == false);
+  if (all.weights.sparse) THROW("--automl does not work with sparse weights");
+
+  details::fail_if_enabled(all,
+      {"ccb_explore_adf", "audit_regressor", "baseline", "cb_explore_adf_rnd", "cb_to_cb_adf", "cbify", "replay_c",
+          "replay_b", "replay_m", "memory_tree", "new_mf", "nn", "stage_poly"});
+
+  // only this has been tested
+  if (base_learner->is_multiline)
+  {
+    // fetch cb_explore_adf to call directly into the print routine twice
+    data->adf_learner = as_multiline(base_learner->get_learner_by_name_prefix("cb_explore_adf_"));
+    auto ppw = max_live_configs;
+
+    auto* l = make_reduction_learner(std::move(data), as_multiline(base_learner),
+        learn_automl<interaction_config_manager, true>, predict_automl<interaction_config_manager, true>,
+        stack_builder.get_setupfn_name(automl_setup))
+                  .set_params_per_weight(ppw)  // refactor pm
+                  .set_finish_example(finish_example<interaction_config_manager>)
+                  .set_save_load(save_load_aml<interaction_config_manager>)
+                  .set_persist_metrics(persist<interaction_config_manager>)
+                  .set_prediction_type(base_learner->pred_type)
+                  .set_label_type(label_type_t::cb)
```

Comment:
Is this a guarantee?

Reply:
This should probably be taken out. As of now automl only support `cb`, but should have a greater scope in the future.